File: app/src/main/java/com/tcssol/expensetracker/Data/PersonExpRepository.java
package com.tcssol.expensetracker.Data;

import android.app.Application;

import androidx.lifecycle.LiveData;

import com.tcssol.expensetracker.Model.Expenses;
import com.tcssol.expensetracker.Model.PersonExp;

import java.util.List;
public class PersonExpRepository {
    private final PersonExpDao personExpDao;
    private final LiveData<List<PersonExp>> allExpenses;
    private final List<PersonExp> allExpensesList;
    private final LiveData<List<PersonExp>> allExpensesGrouped;
    public PersonExpRepository(Application application){
        PersonExpDatabase database=PersonExpDatabase.getDatabase(application);
        this.personExpDao=database.personExpDao();
        this.allExpenses=personExpDao.getExpenses();
        this.allExpensesList=personExpDao.getExpensesList();
        this.allExpensesGrouped= personExpDao.getExpensesGrouped();
    }

    public List<PersonExp> getAllExpensesList() {
        return allExpensesList;
    }
    public LiveData<List<PersonExp>> getAllExpensesGrouped(){
        return allExpensesGrouped;
    }
    public LiveData<List<PersonExp>> getAllExpensesGroupedFiltered(String month,String year){
        if(month.length()==1){
            month="0"+month;
        }
        return personExpDao.getExpensesFilteredGrouped(month,year);
    }
    public LiveData<List<PersonExp>> getAllExpensesFiltered(String month,String year){
        if(month.length()==1){
            month="0"+month;
        }
        return personExpDao.getExpensesFiltered(month,year);
    }

    public LiveData<List<PersonExp>> getAllExpenses(){return allExpenses;}
    public void insert(PersonExp personExp){
        PersonExpDatabase.databaseWriterExecutor.execute(()->personExpDao.insertPersonExpense(personExp));
    }
    public LiveData<PersonExp> get(long id){return personExpDao.get(id);}

    public void update(PersonExp exp){
        PersonExpDatabase.databaseWriterExecutor.execute(()->personExpDao.update(exp));
    }
    public void delete(PersonExp exp){
        PersonExpDatabase.databaseWriterExecutor.execute(()->personExpDao.delete(exp));
    }
    public PersonExpDao getPersonExpDao(){return personExpDao;}

}

File: app/src/main/java/com/tcssol/expensetracker/Data/PersonExpRepository.kt
package com.tcssol.expensetracker.Data

import android.app.Application
import com.tcssol.expensetracker.Model.PersonExp
import kotlinx.coroutines.flow.Flow

class PersonExpRepository(application: Application) {
    private val personExpDao: PersonExpDao = PersonExpDatabase.getDatabase(application).personExpDao()
    val allExpenses: Flow<List<PersonExp>> = personExpDao.getExpenses()
    val allExpensesGrouped: Flow<List<PersonExp>> = personExpDao.getExpensesGrouped()

    suspend fun getAllExpensesList(): List<PersonExp> {
        return personExpDao.getExpensesList()
    }

    suspend fun getAllExpensesListSync(): List<PersonExp> {
        return personExpDao.getExpensesList()
    }

    fun getAllExpensesGroupedFiltered(montht: String, yeart: String): Flow<List<PersonExp>> {
        var month = montht
        if (month.length == 1) month = "0$month"
        return personExpDao.getExpensesFilteredGrouped(month, yeart)
    }

    fun getAllExpensesFiltered(montht: String, yeart: String): Flow<List<PersonExp>> {
        var month = montht
        if (month.length == 1) month = "0$month"
        return personExpDao.getExpensesFiltered(month, yeart)
    }

    suspend fun insert(personExp: PersonExp) {
        personExpDao.insertPersonExpense(personExp)
    }

    fun get(id: Long): Flow<PersonExp?> {
        return personExpDao.get(id)
    }

    suspend fun update(exp: PersonExp) {
        personExpDao.update(exp)
    }

    suspend fun delete(exp: PersonExp) {
        personExpDao.delete(exp)
    }

    fun getPersonExpDao(): PersonExpDao {
        return personExpDao
    }
}
